### PR TITLE
several improvements for next version

### PR DESCRIPTION
**Pull request by alon-e**
improve memory allocation by moving allocation out of converters
add `checkConsistency` API call for transaction promotion
use dynamic snapshots for tip selection
limit inflow of new transactions from single neighbor/user
move address syntax check to txValidation
mark invalid milestones
properly dispose ColumnFamilyHandles when closing DB

### CODE REVIEW

File: src/main/java/com/iota/iri/IRI.java [hunk 1]
```diff
@@ -220,6 +222,10 @@ private static void validateParams(final Configuration configuration, final Stri
         if (vsendLimit != null) {
             configuration.put(DefaultConfSettings.SEND_LIMIT, vsendLimit);
         }
+        if (parser.getOptionValue(sync) != null) {
+            configuration.put(DefaultConfSettings.NEW_TX_LIMIT, "0.0");
```
Comment (th0br0) [inline src/main/java/com/iota/iri/IRI.java]: This should be "9.0" according to the default value further below?

File: src/main/java/com/iota/iri/Milestone.java [hunk 2]
```diff
@@ -84,14 +95,18 @@ public void init(final SpongeFactory.Mode mode, final LedgerValidator ledgerVali
...
-                                        } else {
-                                            analyzedMilestoneCandidates.remove(t.getHash());
+                                        final Validity valid = validateMilestone(mode, t, getIndex(t));
+                                        switch (valid) {
+                                            case VALID:
+                                                MilestoneViewModel milestoneViewModel = MilestoneViewModel.latest(tangle);
+                                                if (milestoneViewModel != null && milestoneViewModel.index() > latestMilestoneIndex) {
+                                                    latestMilestone = milestoneViewModel.getHash();
+                                                    latestMilestoneIndex = milestoneViewModel.index();
+                                                }
```
Comment (th0br0) [inline src/main/java/com/iota/iri/Milestone.java]: aren't you missing breaks here?

File: src/main/java/com/iota/iri/network/Node.java [hunk 3]
```diff
@@ -711,6 +711,17 @@ public Neighbor newNeighbor(final URI uri, boolean isConfigured) {
         return Optional.empty();
     }
 
+    private void parseNeighborsConfig() {
+        Arrays.stream(configuration.string(Configuration.DefaultConfSettings.NEIGHBORS).split(" ")).distinct()
+                .filter(s -> !s.isEmpty()).map(Node::uri).map(Optional::get)
+                .filter(u -> isUriValid(u))
```
Comment (th0br0) [inline src/main/java/com/iota/iri/network/Node.java]: We should log invalid neighbours, could you create an issue?

File: src/main/java/com/iota/iri/service/API.java [hunk 4]
```diff
@@ -318,6 +340,57 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
+        }
+
+        long now = System.currentTimeMillis();
+        if ((now - broadcastStoreTimer.get()) >  Neighbor.newTransactionsWindow) {
+            broadcastStoreCounters.clear();
+            broadcastStoreTimer.set(now);
+        }
+        if(broadcastStoreCounters.putIfAbsent(sourceAddress, new AtomicInteger(size)) != null) {
+            return size < newTransactionsLimit;
+        } else return broadcastStoreCounters.get(sourceAddress).addAndGet(size) < newTransactionsLimit;
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: add braces

File: src/main/java/com/iota/iri/service/API.java [hunk 5]
```diff
@@ -318,6 +340,57 @@ private AbstractResponse process(final String requestString, InetSocketAddress s
...
+    }
+    private AbstractResponse checkConsistencyStatement(List<String> transactionsList) throws Exception {
+        final List<Hash> transactions = transactionsList.stream().map(Hash::new).collect(Collectors.toList());
+        boolean state = true;
+        String info = null;
+
+        //check transactions themselves are valid
+        for (Hash transaction :transactions) {
```
Comment (th0br0) [inline src/main/java/com/iota/iri/service/API.java]: missing space

File: src/main/java/com/iota/iri/IRI.java (hunk 1 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/IRI.java]: this is intended for syncing nodes.
they can start their node w/ `--sync` to bypass the `new_tx_limit`.
(this isn't needed on normal "older" nodes, b/c the syncing node wont send a lot of *new* transactions while syncing)

File: src/main/java/com/iota/iri/network/Node.java (hunk 3 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/network/Node.java]: `isUriValid` logs the invalid entry: [Node.java:689](https://github.com/alon-e/iri/blob/948b8d0b5103903ec5445d474ac62ce3d4d2f6bb/src/main/java/com/iota/iri/network/Node.java#L689)

**Comment by mlouielu** [comment]
Please, break your update into multiple small PR for others to review. Using one HUGE PR with many changed is irresponsible for your user and other developers.

File: src/main/java/com/iota/iri/TransactionValidator.java [hunk 6]
```diff
@@ -75,19 +76,18 @@ private static void runValidation(TransactionViewModel transactionViewModel, fin
...
         if(weightMagnitude < minWeightMagnitude) {
-            /*
-            log.error("Hash found: {}", transactionViewModel.getHash());
-            log.error("Transaction trytes: "+Converter.trytes(transactionViewModel.trits()));
-            */
             throw new RuntimeException("Invalid transaction hash");
         }
+
+        if (transactionViewModel.value() != 0 && transactionViewModel.getAddressHash().trits()[Curl.HASH_LENGTH - 1] != 0) {
+            throw new RuntimeException("Invalid transaction address");
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/TransactionValidator.java]: Why if a address last trits isn't 0 will be invalid?

File: src/main/java/com/iota/iri/conf/Configuration.java [hunk 7]
```diff
@@ -92,6 +93,7 @@
         conf.put(DefaultConfSettings.MAIN_DB.name(), "rocksdb");
         conf.put(DefaultConfSettings.EXPORT.name(), "false");
         conf.put(DefaultConfSettings.SEND_LIMIT.name(), "-1.0");
+        conf.put(DefaultConfSettings.NEW_TX_LIMIT.name(), "9.0");
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/conf/Configuration.java]: What is new tx limit?

File: src/main/java/com/iota/iri/controllers/TipsViewModel.java [hunk 8]
```diff
@@ -1,27 +1,22 @@
...
 
 /**
  * Created by paul on 3/14/17 for iri-testnet.
  */
 public class TipsViewModel {
     final Logger log = LoggerFactory.getLogger(TipsViewModel.class);
 
-    private FifoHashCache tips = new FifoHashCache(5000);
-    private FifoHashCache solidTips = new FifoHashCache(5000);
+    public static final int MAX_TIPS = 5000;
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/controllers/TipsViewModel.java]: Why not configurable?

File: src/main/java/com/iota/iri/controllers/TipsViewModel.java [hunk 9]
```diff
@@ -1,27 +1,22 @@
...
+    private FifoHashCache<Hash> tips = new FifoHashCache<>(TipsViewModel.MAX_TIPS);
+    private FifoHashCache<Hash> solidTips = new FifoHashCache<>(TipsViewModel.MAX_TIPS);
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/controllers/TipsViewModel.java]: Do we need separate configure? (`MAX_SOLID_TIPS`)

File: src/main/java/com/iota/iri/network/Node.java [hunk 10]
```diff
@@ -145,6 +137,8 @@ public void init() throws Exception {
         executor.shutdown();
     }
 
+
+
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/network/Node.java]: three blank line?

File: src/main/java/com/iota/iri/network/Node.java [hunk 11]
```diff
@@ -675,15 +666,12 @@ public void shutdown() throws InterruptedException {
     // helpers methods
 
     public boolean removeNeighbor(final URI uri, boolean isConfigured) {
-        Neighbor neighbor;
-        if (uri.toString().contains("tcp:")) {
-            neighbor = new TCPNeighbor(new InetSocketAddress(uri.getHost(), uri.getPort()), isConfigured);
+        final Neighbor neighbor = newNeighbor(uri,isConfigured);
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/network/Node.java]: whitespace after `,`.

File: src/main/java/com/iota/iri/network/TransactionRequester.java [hunk 12]
```diff
@@ -20,17 +20,17 @@
     private final MessageQ messageQ;
     private final Set<Hash> milestoneTransactionsToRequest = new LinkedHashSet<>();
     private final Set<Hash> transactionsToRequest = new LinkedHashSet<>();
+
+    public static final int MAX_TX_REQ_QUEUE_SIZE = 10000;
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/network/TransactionRequester.java]: configurable?

File: src/main/java/com/iota/iri/Milestone.java [hunk 13]
```diff
@@ -187,33 +205,32 @@ private boolean validateMilestone(SpongeFactory.Mode mode, TransactionViewModel
                                 transactionViewModel2.trits(), 0, index, NUMBER_OF_KEYS_IN_A_MILESTONE);
                         if (testnet || (new Hash(merkleRoot)).equals(coordinator)) {
                             new MilestoneViewModel(index, transactionViewModel.getHash()).store(tangle);
-                            return true;
+                            return VALID;
+                        } else {
+                            //invalid Milestone
```
Comment (mlouielu) [inline src/main/java/com/iota/iri/Milestone.java]: Why need this comment as your return shows it is `INVALID`.

Comment (alon-e) [reply src/main/java/com/iota/iri/Milestone.java]: thx, will remove.

File: src/main/java/com/iota/iri/TransactionValidator.java (hunk 6 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/TransactionValidator.java]: yes,
this is already the case, due to Kerl conversion.
till now this was enforced in BundleValidation: https://github.com/iotaledger/iri/blob/dev/src/main/java/com/iota/iri/BundleValidator.java#L53

this is added to not propagate such transactions in the first place.

File: src/main/java/com/iota/iri/conf/Configuration.java (hunk 7 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/conf/Configuration.java]: this is a limit on the incoming *new* transactions from 1 neighbor,
it's added to artificially simulate network saturation a case where a node is flooded with transactions.

this limit will be set to "off" by default.

File: src/main/java/com/iota/iri/controllers/TipsViewModel.java (hunk 8 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/controllers/TipsViewModel.java]: this class don't have access to configuration.
as configuration will be redone shortly, it's not worth changing the interface imo
(see https://github.com/iotaledger/iri/pull/404)

File: src/main/java/com/iota/iri/controllers/TipsViewModel.java (hunk 9 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/controllers/TipsViewModel.java]: this class don't have access to configuration.
as configuration will be redone shortly, it's not worth changing the interface imo

File: src/main/java/com/iota/iri/network/TransactionRequester.java (hunk 12 above)
Comment (alon-e) [reply src/main/java/com/iota/iri/network/TransactionRequester.java]: this class don't have access to configuration.
as configuration will be redone shortly, it's not worth changing the interface imo

**Comment by PatrickHuetter** [comment]
FindBugs is very good if you don't want external dependencies to some server-side analysis tools. I would also recommend it for the project.

**Pull request merged**